# Table of Contents
1. [Introduction](#id0)
2. [Import Packages](#id1)
3. [Data Import from FARS National Survey](#id2)
4. [Making DataFrames](#id3)
5. [Initial Assessments of Column Types](#id4)
6. [Assessing Possible Joins](#id5)
7. [Checking Columns and Odd Cases](#id6)
8. [Trimming Dataframe to Helpful Columns](#id7)
9. [Saving Data](#id8)



<a id="id0"></a>

## Introduction
Data was drawn from the Fatality Analysis Reporting System (FARS) of the National Highway Traffic Safety Administration (NHTSA). As the FARS manual notes, " Crashes each year result in thousands of lives lost, hundreds of thousands of injured victims, and billions of dollars in property damage. Accurate data are required to support
the development, implementation, and assessment of highway safety programs aimed at reducing
this toll." The FARS data was collected to help improve traveler safety. The FARS data are specific in that " To qualify as a FARS case, the crash had to involve a motor vehicle traveling on a
trafficway customarily open to the public, and must have resulted in the death of a motorist or a
non-motorist within 30 days of the crash."

National Center for Statistics and Analysis. (2022, March). *Fatality Analysis Reporting System
analytical user’s manual*, 1975-2020 (Report No. DOT HS 813 254). National Highway
Traffic Safety Administration. 

<a id="id1"></a>

## Import Packages

In [100]:

import pandas as pd
import csv
import os


<a id="id2"></a>

## Data Import from FARS National Survey

In [101]:
### downloaded 19 CSVs from here https://www.nhtsa.gov/file-downloads?p=nhtsa/downloads/FARS/2019/ 
data_accident = "/Users/kirk/DS/Springboard_DST/Capstone_2/data/external/FARS2019NationalCSV/accident.CSV"
data_person = "/Users/kirk/DS/Springboard_DST/Capstone_2/data/external/FARS2019NationalCSV/Person.CSV"
#data_cevent = "/Users/kirk/DS/Springboard_DST/Capstone_2/data/external/FARS2019NationalCSV/CEvent.CSV"
data_factor = "/Users/kirk/DS/Springboard_DST/Capstone_2/data/external/FARS2019NationalCSV/Factor.CSV"
data_vision = "/Users/kirk/DS/Springboard_DST/Capstone_2/data/external/FARS2019NationalCSV/Vision.csv"
data_drugs = "/Users/kirk/DS/Springboard_DST/Capstone_2/data/external/FARS2019NationalCSV/Drugs.csv"
#data_race = "/Users/kirk/DS/Springboard_DST/Capstone_2/data/external/FARS2019NationalCSV/Race.CSV"

<a id="id3"></a>

## Making DataFrames

In [102]:
table_accident = pd.read_csv(data_accident, low_memory=False)
table_accident.info

<bound method DataFrame.info of        STATE STATENAME  ST_CASE  VE_TOTAL  VE_FORMS  PVH_INVL  PEDS  PERSONS  \
0          1   Alabama    10001         2         2         0     0        3   
1          1   Alabama    10002         2         2         0     0        2   
2          1   Alabama    10003         3         3         0     0        4   
3          1   Alabama    10004         1         1         0     1        1   
4          1   Alabama    10005         1         1         0     0        1   
...      ...       ...      ...       ...       ...       ...   ...      ...   
33482     56   Wyoming   560117         1         1         0     0        1   
33483     56   Wyoming   560118         5         5         0     0        6   
33484     56   Wyoming   560119         1         1         0     0        1   
33485     56   Wyoming   560120         1         1         0     0        2   
33486     56   Wyoming   560121         2         2         0     0        3   

       

In [103]:
table_accident.columns

Index(['STATE', 'STATENAME', 'ST_CASE', 'VE_TOTAL', 'VE_FORMS', 'PVH_INVL',
       'PEDS', 'PERSONS', 'PERMVIT', 'PERNOTMVIT', 'COUNTY', 'COUNTYNAME',
       'CITY', 'CITYNAME', 'DAY', 'DAYNAME', 'MONTH', 'MONTHNAME', 'YEAR',
       'DAY_WEEK', 'DAY_WEEKNAME', 'HOUR', 'HOURNAME', 'MINUTE', 'MINUTENAME',
       'NHS', 'NHSNAME', 'ROUTE', 'ROUTENAME', 'TWAY_ID', 'TWAY_ID2',
       'RUR_URB', 'RUR_URBNAME', 'FUNC_SYS', 'FUNC_SYSNAME', 'RD_OWNER',
       'RD_OWNERNAME', 'MILEPT', 'MILEPTNAME', 'LATITUDE', 'LATITUDENAME',
       'LONGITUD', 'LONGITUDNAME', 'SP_JUR', 'SP_JURNAME', 'HARM_EV',
       'HARM_EVNAME', 'MAN_COLL', 'MAN_COLLNAME', 'RELJCT1', 'RELJCT1NAME',
       'RELJCT2', 'RELJCT2NAME', 'TYP_INT', 'TYP_INTNAME', 'WRK_ZONE',
       'WRK_ZONENAME', 'REL_ROAD', 'REL_ROADNAME', 'LGT_COND', 'LGT_CONDNAME',
       'WEATHER1', 'WEATHER1NAME', 'WEATHER2', 'WEATHER2NAME', 'WEATHER',
       'WEATHERNAME', 'SCH_BUS', 'SCH_BUSNAME', 'RAIL', 'RAILNAME', 'NOT_HOUR',
       'NOT_HOURNAME', 'NOT

In [104]:
table_accident['MONTHNAME'].value_counts()

September    3093
August       3091
July         3037
October      2972
June         2926
May          2917
November     2841
December     2774
April        2611
March        2544
January      2476
February     2205
Name: MONTHNAME, dtype: int64

In [105]:
table_accident.FATALS.value_counts()

1    31123
2     1986
3      287
4       70
5       12
6        5
7        3
8        1
Name: FATALS, dtype: int64

In [106]:
table_person = pd.read_csv(data_person, low_memory=False)
table_person.head(5)

STATE STATENAME  ST_CASE  VE_FORMS  VEH_NO  PER_NO  STR_VEH  COUNTY  DAY  \
0      1   Alabama    10001         2       1       1        0      81    7   
1      1   Alabama    10001         2       1       2        0      81    7   
2      1   Alabama    10001         2       2       1        0      81    7   
3      1   Alabama    10002         2       1       1        0      55   23   
4      1   Alabama    10002         2       2       1        0      55   23   

   DAYNAME  ...  WORK_INJ                     WORK_INJNAME  HISPANIC  \
0        7  ...         8  Not Applicable (not a fatality)         0   
1        7  ...         0                               No         7   
2        7  ...         8  Not Applicable (not a fatality)         0   
3       23  ...         0                               No         7   
4       23  ...         8  Not Applicable (not a fatality)         0   

                      HISPANICNAME  LOCATION                 LOCATIONNAME  \
0  Not A Fatality (not Applicable)         0  Occupant of a Motor Vehicle   
1                     Non-Hispanic         0  Occupant of a Motor Vehicle   
2  Not A Fatality (not Applicable)         0  Occupant of a Motor Vehicle   
3                     Non-Hispanic         0  Occupant of a Motor Vehicle   
4  Not A Fatality (not Applicable)         0  Occupant of a Motor Vehicle   

   HELM_USE    HELM_USENAME  HELM_MIS              HELM_MISNAME  
0        20  Not Applicable         7  None Used/Not Applicable  
1        20  Not Applicable         7  None Used/Not Applicable  
2        20  Not Applicable         7  None Used/Not Applicable  
3        20  Not Applicable         7  None Used/Not Applicable  
4        20  Not Applicable         7  None Used/Not Applicable  

[5 rows x 119 columns]

In [107]:
pd.unique(table_person.columns)

array(['STATE', 'STATENAME', 'ST_CASE', 'VE_FORMS', 'VEH_NO', 'PER_NO',
       'STR_VEH', 'COUNTY', 'DAY', 'DAYNAME', 'MONTH', 'MONTHNAME',
       'HOUR', 'HOURNAME', 'MINUTE', 'MINUTENAME', 'RUR_URB',
       'RUR_URBNAME', 'FUNC_SYS', 'FUNC_SYSNAME', 'HARM_EV',
       'HARM_EVNAME', 'MAN_COLL', 'MAN_COLLNAME', 'SCH_BUS',
       'SCH_BUSNAME', 'MAKE', 'MAKENAME', 'MAK_MOD', 'BODY_TYP',
       'BODY_TYPNAME', 'MOD_YEAR', 'MOD_YEARNAME', 'TOW_VEH',
       'TOW_VEHNAME', 'SPEC_USE', 'SPEC_USENAME', 'EMER_USE',
       'EMER_USENAME', 'ROLLOVER', 'ROLLOVERNAME', 'IMPACT1',
       'IMPACT1NAME', 'FIRE_EXP', 'FIRE_EXPNAME', 'AGE', 'AGENAME', 'SEX',
       'SEXNAME', 'PER_TYP', 'PER_TYPNAME', 'INJ_SEV', 'INJ_SEVNAME',
       'SEAT_POS', 'SEAT_POSNAME', 'REST_USE', 'REST_USENAME', 'REST_MIS',
       'REST_MISNAME', 'AIR_BAG', 'AIR_BAGNAME', 'EJECTION',
       'EJECTIONNAME', 'EJ_PATH', 'EJ_PATHNAME', 'EXTRICAT',
       'EXTRICATNAME', 'DRINKING', 'DRINKINGNAME', 'ALC_DET',
       'ALC_DETNAME',

In [108]:
#### assessing values in PERSON to decide which to keep
# pd.unique(table_person.DOANAME)
table_person.DOANAME.value_counts()

Not Applicable    62413
Died at Scene     20121
Died En Route       241
Unknown              68
Name: DOANAME, dtype: int64

In [109]:
### Need to craft a fatality by person column

In [110]:
table_factor = pd.read_csv(data_factor, low_memory=False)
table_factor.head()

STATE STATENAME  ST_CASE  VEH_NO  MFACTOR MFACTORNAME
0      1   Alabama    10001       1        0        None
1      1   Alabama    10001       2        0        None
2      1   Alabama    10002       1        0        None
3      1   Alabama    10002       2        0        None
4      1   Alabama    10003       1        0        None

In [111]:
### checking out factor names in factors
pd.unique(table_factor['MFACTORNAME'])

array(['None', 'Reported as Unknown',
       'Vehicle Contributing Factors - No Details', 'Tires',
       'Other Lights', 'Other', 'Brake System', 'Head Lights',
       'Not Reported', 'Steering', 'Windows/Windshield',
       'Truck Coupling/Trailer Hitch/Safety Chains', 'Suspension',
       'Safety Systems', 'Body, Doors', 'Wheels', 'Mirrors',
       'Power Train', 'Exhaust System', 'Signal Lights', 'Wipers'],
      dtype=object)

In [112]:
table_vision = pd.read_csv(data_vision, low_memory=False)
table_vision.head()

STATE STATENAME  ST_CASE  VEH_NO  MVISOBSC          MVISOBSCNAME
0      1   Alabama    10001       1         0  No Obstruction Noted
1      1   Alabama    10001       2         0  No Obstruction Noted
2      1   Alabama    10002       1         0  No Obstruction Noted
3      1   Alabama    10002       2         0  No Obstruction Noted
4      1   Alabama    10003       1         0  No Obstruction Noted

In [113]:
### checking out factor names in vision
pd.unique(table_vision['MVISOBSCNAME'])

array(['No Obstruction Noted',
       'Curve, Hill or Other Roadway Design Feature',
       'No Driver Present/Unknown if Driver present',
       'Reflected Glare, Bright Sunlight, Headlights',
       'In-Transport Motor Vehicle (including load)',
       'Rain, Snow, Fog, Smoke, Sand, Dust', 'Reported as Unknown',
       'Vision Obscured - No Details',
       'Not In-Transport Motor Vehicle (parked, working)',
       'Other Visual Obstruction', 'Inadequate Defrost or Defog System',
       'Trees, Crops, Vegetation', 'Splash or Spray of Passing Vehicle',
       'Obstruction Interior to the Vehicle',
       'Obstructing Angles on Vehicle',
       'Inadequate Vehicle Lighting System', 'External Mirrors',
       'Broken or Improperly Cleaned Windshield',
       'Building, Billboard, Other Structure'], dtype=object)

<a id="id2"></a>

In [114]:
table_drugs = pd.read_csv(data_drugs, low_memory=False)
table_drugs.head()

STATE STATENAME  ST_CASE  VEH_NO  PER_NO  DRUGSPEC    DRUGSPECNAME  \
0      1   Alabama    10001       1       1         0  Test Not Given   
1      1   Alabama    10001       1       2         0  Test Not Given   
2      1   Alabama    10001       2       1         0  Test Not Given   
3      1   Alabama    10002       1       1         1     Whole Blood   
4      1   Alabama    10002       1       1         1     Whole Blood   

   DRUGRES      DRUGRESNAME  
0        0   Test Not Given  
1        0   Test Not Given  
2        0   Test Not Given  
3      401      AMPHETAMINE  
4      417  METHAMPHETAMINE

<a id="id4"></a>

## Initial Assessments of Column Types

In [115]:
### Data wrangling challenge will be to see how and if the columns from different DataFrames should be combined
table_accident.columns

Index(['STATE', 'STATENAME', 'ST_CASE', 'VE_TOTAL', 'VE_FORMS', 'PVH_INVL',
       'PEDS', 'PERSONS', 'PERMVIT', 'PERNOTMVIT', 'COUNTY', 'COUNTYNAME',
       'CITY', 'CITYNAME', 'DAY', 'DAYNAME', 'MONTH', 'MONTHNAME', 'YEAR',
       'DAY_WEEK', 'DAY_WEEKNAME', 'HOUR', 'HOURNAME', 'MINUTE', 'MINUTENAME',
       'NHS', 'NHSNAME', 'ROUTE', 'ROUTENAME', 'TWAY_ID', 'TWAY_ID2',
       'RUR_URB', 'RUR_URBNAME', 'FUNC_SYS', 'FUNC_SYSNAME', 'RD_OWNER',
       'RD_OWNERNAME', 'MILEPT', 'MILEPTNAME', 'LATITUDE', 'LATITUDENAME',
       'LONGITUD', 'LONGITUDNAME', 'SP_JUR', 'SP_JURNAME', 'HARM_EV',
       'HARM_EVNAME', 'MAN_COLL', 'MAN_COLLNAME', 'RELJCT1', 'RELJCT1NAME',
       'RELJCT2', 'RELJCT2NAME', 'TYP_INT', 'TYP_INTNAME', 'WRK_ZONE',
       'WRK_ZONENAME', 'REL_ROAD', 'REL_ROADNAME', 'LGT_COND', 'LGT_CONDNAME',
       'WEATHER1', 'WEATHER1NAME', 'WEATHER2', 'WEATHER2NAME', 'WEATHER',
       'WEATHERNAME', 'SCH_BUS', 'SCH_BUSNAME', 'RAIL', 'RAILNAME', 'NOT_HOUR',
       'NOT_HOURNAME', 'NOT

In [116]:
table_accident['MONTH'].head()

0    2
1    1
2    1
3    1
4    1
Name: MONTH, dtype: int64

In [117]:
### Really large column list for person data
print(table_person.columns.tolist())

['STATE', 'STATENAME', 'ST_CASE', 'VE_FORMS', 'VEH_NO', 'PER_NO', 'STR_VEH', 'COUNTY', 'DAY', 'DAYNAME', 'MONTH', 'MONTHNAME', 'HOUR', 'HOURNAME', 'MINUTE', 'MINUTENAME', 'RUR_URB', 'RUR_URBNAME', 'FUNC_SYS', 'FUNC_SYSNAME', 'HARM_EV', 'HARM_EVNAME', 'MAN_COLL', 'MAN_COLLNAME', 'SCH_BUS', 'SCH_BUSNAME', 'MAKE', 'MAKENAME', 'MAK_MOD', 'BODY_TYP', 'BODY_TYPNAME', 'MOD_YEAR', 'MOD_YEARNAME', 'TOW_VEH', 'TOW_VEHNAME', 'SPEC_USE', 'SPEC_USENAME', 'EMER_USE', 'EMER_USENAME', 'ROLLOVER', 'ROLLOVERNAME', 'IMPACT1', 'IMPACT1NAME', 'FIRE_EXP', 'FIRE_EXPNAME', 'AGE', 'AGENAME', 'SEX', 'SEXNAME', 'PER_TYP', 'PER_TYPNAME', 'INJ_SEV', 'INJ_SEVNAME', 'SEAT_POS', 'SEAT_POSNAME', 'REST_USE', 'REST_USENAME', 'REST_MIS', 'REST_MISNAME', 'AIR_BAG', 'AIR_BAGNAME', 'EJECTION', 'EJECTIONNAME', 'EJ_PATH', 'EJ_PATHNAME', 'EXTRICAT', 'EXTRICATNAME', 'DRINKING', 'DRINKINGNAME', 'ALC_DET', 'ALC_DETNAME', 'ALC_STATUS', 'ALC_STATUSNAME', 'ATST_TYP', 'ATST_TYPNAME', 'ALC_RES', 'ALC_RESNAME', 'DRUGS', 'DRUGSNAME', 'D

In [118]:
#### Need to subset for important columns only in accident, winnow the other tables, and then combine
df_accident = table_accident[['STATENAME', 'ST_CASE','PEDS', 'RUR_URBNAME', 'HARM_EVNAME', 
                             'TYP_INTNAME', 'LGT_CONDNAME', 'WEATHER1NAME', 'WEATHER2NAME', 
                              'WEATHERNAME','FATALS','DRUNK_DR','DAY_WEEKNAME' , 'HOUR', 'MONTHNAME']]    

df_person = table_person[['STATE', 'ST_CASE', 'VE_FORMS', 'VEH_NO', 'PER_NO', 'AGE', 'SEXNAME','PER_TYPNAME','DRINKINGNAME',]]

df_factor = table_factor[['ST_CASE', 'VEH_NO', 'MFACTORNAME']]

df_vision = table_vision[['ST_CASE', 'VEH_NO','MVISOBSCNAME']]

df_drugs = table_drugs[['ST_CASE', 'VEH_NO', 'PER_NO', 'DRUGRESNAME']]


In [119]:
df_accident.columns

Index(['STATENAME', 'ST_CASE', 'PEDS', 'RUR_URBNAME', 'HARM_EVNAME',
       'TYP_INTNAME', 'LGT_CONDNAME', 'WEATHER1NAME', 'WEATHER2NAME',
       'WEATHERNAME', 'FATALS', 'DRUNK_DR', 'DAY_WEEKNAME', 'HOUR',
       'MONTHNAME'],
      dtype='object')

In [120]:
df_accident.head()

STATENAME  ST_CASE  PEDS RUR_URBNAME                 HARM_EVNAME  \
0   Alabama    10001     0       Urban  Motor Vehicle In-Transport   
1   Alabama    10002     0       Urban  Motor Vehicle In-Transport   
2   Alabama    10003     0       Rural  Motor Vehicle In-Transport   
3   Alabama    10004     1       Rural                  Pedestrian   
4   Alabama    10005     0       Urban           Rollover/Overturn   

           TYP_INTNAME        LGT_CONDNAME      WEATHER1NAME  \
0  Not an Intersection            Daylight             Clear   
1  Not an Intersection  Dark - Not Lighted              Rain   
2  Not an Intersection  Dark - Not Lighted            Cloudy   
3  Not an Intersection  Dark - Not Lighted             Clear   
4  Not an Intersection  Dark - Not Lighted  Fog, Smog, Smoke   

                           WEATHER2NAME       WEATHERNAME  FATALS  DRUNK_DR  \
0  No Additional Atmospheric Conditions             Clear       1         1   
1  No Additional Atmospheric Conditions              Rain       1         0   
2  No Additional Atmospheric Conditions            Cloudy       1         0   
3  No Additional Atmospheric Conditions             Clear       1         0   
4  No Additional Atmospheric Conditions  Fog, Smog, Smoke       1         1   

  DAY_WEEKNAME  HOUR MONTHNAME  
0     Thursday    12  February  
1    Wednesday    18   January  
2      Tuesday    19   January  
3      Tuesday     3   January  
4       Friday     5   January

In [121]:
df_person.columns

Index(['STATE', 'ST_CASE', 'VE_FORMS', 'VEH_NO', 'PER_NO', 'AGE', 'SEXNAME',
       'PER_TYPNAME', 'DRINKINGNAME'],
      dtype='object')

In [122]:
df_factor.columns

Index(['ST_CASE', 'VEH_NO', 'MFACTORNAME'], dtype='object')

In [123]:
df_vision.columns

Index(['ST_CASE', 'VEH_NO', 'MVISOBSCNAME'], dtype='object')

In [124]:
df_drugs.columns

Index(['ST_CASE', 'VEH_NO', 'PER_NO', 'DRUGRESNAME'], dtype='object')

In [125]:
### But drugs names data are too complex for these purposes; let's hold out on simplifing to drugs or no_drugs
df_drugs['DRUGRESNAME'].value_counts()


Test Not Given                     48439
Tested, No Drugs Found/Negative    14313
Not Reported                        6414
Other Drug                          6209
Tetrahydrocannabinols (THC)         3296
                                   ...  
Cloxazolam                             1
Hydroxypethidine                       1
Methylone                              1
ETHYLMORPHINE                          1
Methenolone                            1
Name: DRUGRESNAME, Length: 133, dtype: int64

<a id="id5"></a>

## Assessing Possible Joins

The different dataframes were assessed to see in what useful ways they could be combined to better understand fatalities. However, the dataframes were established as separate tables to provide insights without duplicating state case unique IDs and thereby fatalities. These complications are illustrated below. 


In [126]:
### Adding factor data
df_merged_accident = df_accident.merge(df_factor, on=['ST_CASE'], how = 'left')
df_merged_accident.head(5)

STATENAME  ST_CASE  PEDS RUR_URBNAME                 HARM_EVNAME  \
0   Alabama    10001     0       Urban  Motor Vehicle In-Transport   
1   Alabama    10001     0       Urban  Motor Vehicle In-Transport   
2   Alabama    10002     0       Urban  Motor Vehicle In-Transport   
3   Alabama    10002     0       Urban  Motor Vehicle In-Transport   
4   Alabama    10003     0       Rural  Motor Vehicle In-Transport   

           TYP_INTNAME        LGT_CONDNAME WEATHER1NAME  \
0  Not an Intersection            Daylight        Clear   
1  Not an Intersection            Daylight        Clear   
2  Not an Intersection  Dark - Not Lighted         Rain   
3  Not an Intersection  Dark - Not Lighted         Rain   
4  Not an Intersection  Dark - Not Lighted       Cloudy   

                           WEATHER2NAME WEATHERNAME  FATALS  DRUNK_DR  \
0  No Additional Atmospheric Conditions       Clear       1         1   
1  No Additional Atmospheric Conditions       Clear       1         1   
2  No Additional Atmospheric Conditions        Rain       1         0   
3  No Additional Atmospheric Conditions        Rain       1         0   
4  No Additional Atmospheric Conditions      Cloudy       1         0   

  DAY_WEEKNAME  HOUR MONTHNAME  VEH_NO MFACTORNAME  
0     Thursday    12  February       1        None  
1     Thursday    12  February       2        None  
2    Wednesday    18   January       1        None  
3    Wednesday    18   January       2        None  
4      Tuesday    19   January       1        None

In the cell above, the FATALS for ST_CASE 10001 has been doubled, introducing an error to the data.

In [127]:
#### Considering the rare case of 8 fatalities below, we need to keep the fatalities from duplicating when joining.
df_accident.FATALS.describe()

count    33487.000000
mean         1.085645
std          0.345290
min          1.000000
25%          1.000000
50%          1.000000
75%          1.000000
max          8.000000
Name: FATALS, dtype: float64

In [128]:
### Adding vision
df_merged_accident = df_merged_accident.merge(df_vision, on=['ST_CASE', 'VEH_NO',], how = 'left')
df_merged_accident.head(10)


STATENAME  ST_CASE  PEDS RUR_URBNAME                 HARM_EVNAME  \
0   Alabama    10001     0       Urban  Motor Vehicle In-Transport   
1   Alabama    10001     0       Urban  Motor Vehicle In-Transport   
2   Alabama    10002     0       Urban  Motor Vehicle In-Transport   
3   Alabama    10002     0       Urban  Motor Vehicle In-Transport   
4   Alabama    10003     0       Rural  Motor Vehicle In-Transport   
5   Alabama    10003     0       Rural  Motor Vehicle In-Transport   
6   Alabama    10003     0       Rural  Motor Vehicle In-Transport   
7   Alabama    10004     1       Rural                  Pedestrian   
8   Alabama    10005     0       Urban           Rollover/Overturn   
9   Alabama    10006     0       Rural  Motor Vehicle In-Transport   

           TYP_INTNAME        LGT_CONDNAME      WEATHER1NAME  \
0  Not an Intersection            Daylight             Clear   
1  Not an Intersection            Daylight             Clear   
2  Not an Intersection  Dark - Not Lighted              Rain   
3  Not an Intersection  Dark - Not Lighted              Rain   
4  Not an Intersection  Dark - Not Lighted            Cloudy   
5  Not an Intersection  Dark - Not Lighted            Cloudy   
6  Not an Intersection  Dark - Not Lighted            Cloudy   
7  Not an Intersection  Dark - Not Lighted             Clear   
8  Not an Intersection  Dark - Not Lighted  Fog, Smog, Smoke   
9  Not an Intersection            Daylight             Clear   

                           WEATHER2NAME       WEATHERNAME  FATALS  DRUNK_DR  \
0  No Additional Atmospheric Conditions             Clear       1         1   
1  No Additional Atmospheric Conditions             Clear       1         1   
2  No Additional Atmospheric Conditions              Rain       1         0   
3  No Additional Atmospheric Conditions              Rain       1         0   
4  No Additional Atmospheric Conditions            Cloudy       1         0   
5  No Additional Atmospheric Conditions            Cloudy       1         0   
6  No Additional Atmospheric Conditions            Cloudy       1         0   
7  No Additional Atmospheric Conditions             Clear       1         0   
8  No Additional Atmospheric Conditions  Fog, Smog, Smoke       1         1   
9  No Additional Atmospheric Conditions             Clear       1         0   

  DAY_WEEKNAME  HOUR MONTHNAME  VEH_NO MFACTORNAME          MVISOBSCNAME  
0     Thursday    12  February       1        None  No Obstruction Noted  
1     Thursday    12  February       2        None  No Obstruction Noted  
2    Wednesday    18   January       1        None  No Obstruction Noted  
3    Wednesday    18   January       2        None  No Obstruction Noted  
4      Tuesday    19   January       1        None  No Obstruction Noted  
5      Tuesday    19   January       2        None  No Obstruction Noted  
6      Tuesday    19   January       3        None  No Obstruction Noted  
7      Tuesday     3   January       1        None  No Obstruction Noted  
8       Friday     5   January       1        None  No Obstruction Noted  
9       Monday    12   January       1        None  No Obstruction Noted

In [129]:
df_merged_accident.HOUR.describe().T

count    51735.000000
mean        13.447492
std          9.013299
min          0.000000
25%          8.000000
50%         14.000000
75%         18.000000
max         99.000000
Name: HOUR, dtype: float64

In [130]:
### Adding df_merged_person to df_merged_accident to assess shape
df_merged_accident = df_merged_accident.merge(df_person, on=['ST_CASE', 'VEH_NO'], how = 'left')
df_merged_accident.head(5)


STATENAME  ST_CASE  PEDS RUR_URBNAME                 HARM_EVNAME  \
0   Alabama    10001     0       Urban  Motor Vehicle In-Transport   
1   Alabama    10001     0       Urban  Motor Vehicle In-Transport   
2   Alabama    10001     0       Urban  Motor Vehicle In-Transport   
3   Alabama    10002     0       Urban  Motor Vehicle In-Transport   
4   Alabama    10002     0       Urban  Motor Vehicle In-Transport   

           TYP_INTNAME        LGT_CONDNAME WEATHER1NAME  \
0  Not an Intersection            Daylight        Clear   
1  Not an Intersection            Daylight        Clear   
2  Not an Intersection            Daylight        Clear   
3  Not an Intersection  Dark - Not Lighted         Rain   
4  Not an Intersection  Dark - Not Lighted         Rain   

                           WEATHER2NAME WEATHERNAME  ...  VEH_NO  MFACTORNAME  \
0  No Additional Atmospheric Conditions       Clear  ...       1         None   
1  No Additional Atmospheric Conditions       Clear  ...       1         None   
2  No Additional Atmospheric Conditions       Clear  ...       2         None   
3  No Additional Atmospheric Conditions        Rain  ...       1         None   
4  No Additional Atmospheric Conditions        Rain  ...       2         None   

           MVISOBSCNAME  STATE VE_FORMS  PER_NO   AGE SEXNAME  \
0  No Obstruction Noted    1.0      2.0     1.0  34.0  Female   
1  No Obstruction Noted    1.0      2.0     2.0  53.0    Male   
2  No Obstruction Noted    1.0      2.0     1.0  59.0    Male   
3  No Obstruction Noted    1.0      2.0     1.0  42.0  Female   
4  No Obstruction Noted    1.0      2.0     1.0  54.0  Female   

                                 PER_TYPNAME               DRINKINGNAME  
0     Driver of a Motor Vehicle In-Transport  No (Alcohol Not Involved)  
1  Passenger of a Motor Vehicle In-Transport               Not Reported  
2     Driver of a Motor Vehicle In-Transport     Yes (Alcohol Involved)  
3     Driver of a Motor Vehicle In-Transport        Reported as Unknown  
4     Driver of a Motor Vehicle In-Transport  No (Alcohol Not Involved)  

[5 rows x 25 columns]

Because of these complications, the decision was made to work with the complete accident table until at least after the EDA as it does contain basic information number of persons in the vehicles, drunk driving, number of pedestrians, weather, day, and rural/urban setting for the accident itself.

<a id="id6"></a>

## Checking Columns and Odd Cases

In [131]:
### changing some int64 into categories
table_accident['STATE'] = table_accident['STATE'].astype('category')
table_accident['ST_CASE'] = table_accident['ST_CASE'].astype('category')
table_accident['STATENAME'] = table_accident['STATENAME'].astype('category')
table_accident['DRUNK_DR'] = table_accident['DRUNK_DR'].astype('category')
table_accident['RUR_URBNAME'] = table_accident['RUR_URBNAME'].astype('category')
total = table_accident

In [132]:
total.STATENAME.value_counts()

California              3427
Texas                   3296
Florida                 2952
Georgia                 1378
North Carolina          1358
Tennessee               1041
Ohio                    1039
Pennsylvania             990
Illinois                 938
South Carolina           927
Arizona                  908
Michigan                 903
New York                 879
Alabama                  856
Missouri                 819
Virginia                 774
Indiana                  752
Louisiana                681
Kentucky                 667
Oklahoma                 584
Mississippi              580
Colorado                 545
Wisconsin                527
New Jersey               524
Washington               513
Maryland                 496
Arkansas                 473
Oregon                   455
New Mexico               369
Kansas                   361
Minnesota                333
Massachusetts            323
Iowa                     313
Nevada                   285
West Virginia 

It seems there was a crash in PA where 59 vehicles were involved and there were 2 fatalities; blowing snow with a jackknifed truck. This crash is an outlier, and although it is important for multicar crashes, it will really throw off the axes in any EDA plots. Will also drop all rows where total vehicles is 1.

In [159]:
rare_crash = total[total.VE_TOTAL == 28]
print(rare_crash)

     STATE STATENAME ST_CASE  VE_TOTAL  VE_FORMS  PVH_INVL  PEDS  PERSONS  \
5850     8  Colorado   80125        28        28         0     0       31   

      PERMVIT  PERNOTMVIT  ...  HOSP_MN HOSP_MNNAME  CF1  \
5850       31           0  ...       34          34   19   

                                 CF1NAME  CF2                    CF2NAME  CF3  \
5850  Recent/Previous Crash scene Nearby   27  Backup Due to Prior Crash    0   

     CF3NAME  FATALS  DRUNK_DR  
5850    None       4         0  

[1 rows x 91 columns]


In [160]:
total = total.drop([5850])

In [161]:
total.VE_TOTAL.value_counts()


1     18332
2     12287
3      2072
4       484
5       184
6        61
7        35
8        14
9         5
12        4
11        2
16        2
Name: VE_TOTAL, dtype: int64

In [162]:
### State Case 280246 is an illustrative example of the data. Two vehicle accident where vehicle 1 had 9 people and 
### vehicle 2 had 1. Each row is listed as 8 fatalities, because for THAT ST_CASE number, there were 8. We acutally 
### don't know which of the 9 people in the first vehicle died.

rare_deaths = total[total.FATALS == 8]
rare_deaths.head()

STATE    STATENAME ST_CASE  VE_TOTAL  VE_FORMS  PVH_INVL  PEDS  PERSONS  \
17431    28  Mississippi  280246         2         2         0     0       10   

       PERMVIT  PERNOTMVIT  ...  HOSP_MN                        HOSP_MNNAME  \
17431       10           0  ...       99  Unknown EMS Hospital Arrival Time   

       CF1 CF1NAME  CF2  CF2NAME  CF3 CF3NAME  FATALS  DRUNK_DR  
17431    0    None    0     None    0    None       8         0  

[1 rows x 91 columns]

In [163]:
total.groupby("STATENAME").FATALS.value_counts(sort=True)

STATENAME  FATALS
Alabama    1         799
           2          44
           3          10
           4           2
           5           1
                    ... 
Wisconsin  3           2
           4           1
Wyoming    1          97
           2          19
           3           4
Name: FATALS, Length: 197, dtype: int64

In the EDA, we can explore patterns in the rural and urban datasets themselves.

<a id="id7"></a>

## Trimming Dataframe to Helpful Columns

In [164]:
total.iloc[:,19:24].head()

DAY_WEEK DAY_WEEKNAME  HOUR         HOURNAME  MINUTE
0         5     Thursday    12  12:00pm-12:59pm      54
1         4    Wednesday    18    6:00pm-6:59pm       3
2         3      Tuesday    19    7:00pm-7:59pm       0
3         3      Tuesday     3    3:00am-3:59am      15
4         6       Friday     5    5:00am-5:59am      50

In [165]:
total.columns

Index(['STATE', 'STATENAME', 'ST_CASE', 'VE_TOTAL', 'VE_FORMS', 'PVH_INVL',
       'PEDS', 'PERSONS', 'PERMVIT', 'PERNOTMVIT', 'COUNTY', 'COUNTYNAME',
       'CITY', 'CITYNAME', 'DAY', 'DAYNAME', 'MONTH', 'MONTHNAME', 'YEAR',
       'DAY_WEEK', 'DAY_WEEKNAME', 'HOUR', 'HOURNAME', 'MINUTE', 'MINUTENAME',
       'NHS', 'NHSNAME', 'ROUTE', 'ROUTENAME', 'TWAY_ID', 'TWAY_ID2',
       'RUR_URB', 'RUR_URBNAME', 'FUNC_SYS', 'FUNC_SYSNAME', 'RD_OWNER',
       'RD_OWNERNAME', 'MILEPT', 'MILEPTNAME', 'LATITUDE', 'LATITUDENAME',
       'LONGITUD', 'LONGITUDNAME', 'SP_JUR', 'SP_JURNAME', 'HARM_EV',
       'HARM_EVNAME', 'MAN_COLL', 'MAN_COLLNAME', 'RELJCT1', 'RELJCT1NAME',
       'RELJCT2', 'RELJCT2NAME', 'TYP_INT', 'TYP_INTNAME', 'WRK_ZONE',
       'WRK_ZONENAME', 'REL_ROAD', 'REL_ROADNAME', 'LGT_COND', 'LGT_CONDNAME',
       'WEATHER1', 'WEATHER1NAME', 'WEATHER2', 'WEATHER2NAME', 'WEATHER',
       'WEATHERNAME', 'SCH_BUS', 'SCH_BUSNAME', 'RAIL', 'RAILNAME', 'NOT_HOUR',
       'NOT_HOURNAME', 'NOT

In [166]:
accident = total[['STATENAME','VE_TOTAL','PEDS', 'PERSONS','DAYNAME','MONTHNAME','HOUR',
                  'DAY_WEEKNAME','RUR_URBNAME','HARM_EVNAME','LGT_CONDNAME', 'WEATHER1NAME',
                 'WEATHERNAME', 'FATALS', 'DRUNK_DR']]
accident.reset_index()
accident.head(5)

STATENAME  VE_TOTAL  PEDS  PERSONS  DAYNAME MONTHNAME  HOUR DAY_WEEKNAME  \
0   Alabama         2     0        3        7  February    12     Thursday   
1   Alabama         2     0        2       23   January    18    Wednesday   
2   Alabama         3     0        4       22   January    19      Tuesday   
3   Alabama         1     1        1       22   January     3      Tuesday   
4   Alabama         1     0        1       18   January     5       Friday   

  RUR_URBNAME                 HARM_EVNAME        LGT_CONDNAME  \
0       Urban  Motor Vehicle In-Transport            Daylight   
1       Urban  Motor Vehicle In-Transport  Dark - Not Lighted   
2       Rural  Motor Vehicle In-Transport  Dark - Not Lighted   
3       Rural                  Pedestrian  Dark - Not Lighted   
4       Urban           Rollover/Overturn  Dark - Not Lighted   

       WEATHER1NAME       WEATHERNAME  FATALS DRUNK_DR  
0             Clear             Clear       1        1  
1              Rain              Rain       1        0  
2            Cloudy            Cloudy       1        0  
3             Clear             Clear       1        0  
4  Fog, Smog, Smoke  Fog, Smog, Smoke       1        1

In [167]:
accident.shape

(33482, 15)

In [168]:
## Possibly univariate analysis, the rural data will be split from the urban data
rural = accident[accident['RUR_URBNAME'] == 'Rural']
rural.reset_index()
rural.shape

(14624, 15)

In [169]:
urban = accident[accident['RUR_URBNAME'] == 'Urban']
urban.reset_index()
urban.shape

(18750, 15)

<a id="id8"></a>

## Saving Data

In [172]:
# save the data to a new csv file
accident.to_csv('../data/processed/accident.csv', index=False)
urban.to_csv('../data/processed/urban.csv', index=False)
rural.to_csv('../data/processed/rural.csv', index=False)
